File: socket/src/entity/Thread.ts
import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  UpdateDateColumn,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { Keyword } from './Keyword';
import { User } from './User';

@Index('FKedurjn88vojgqvkjwbxlukmfx', ['authorId'], {})
@Index('FK50e291c4mexo6p78mi3t2xd6u', ['keywordId'], {})
@Index('FKfspfoq62r3wrwmx755muuqnf8', ['parentThreadId'], {})
@Entity('thread')
export class Thread {
  @PrimaryGeneratedColumn('increment', { name: 'id', type: 'bigint' })
  id: string;

  @Column('varchar', { name: 'content', nullable: true, length: 5000 })
  content: string | null;

  @Column('bigint', { name: 'author_id', nullable: true })
  authorId: string | null;

  @Column('bigint', { name: 'keyword_id', nullable: true })
  keywordId: string | null;

  @Column('bigint', { name: 'parent_thread_id', nullable: true })
  parentThreadId: string | null;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @DeleteDateColumn({ name: 'deleted_at' })
  deletedAt: Date | null;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date | null;

  @ManyToOne(() => Keyword, (keyword) => keyword.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'keyword_id', referencedColumnName: 'id' }])
  keyword: Keyword;

  @ManyToOne(() => User, (user) => user.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'author_id', referencedColumnName: 'id' }])
  author: User;

  @ManyToOne(() => Thread, (thread) => thread.threads, {
    onDelete: 'NO ACTION',
    onUpdate: 'NO ACTION',
  })
  @JoinColumn([{ name: 'parent_thread_id', referencedColumnName: 'id' }])
  parentThread: Thread;

  @OneToMany(() => Thread, (thread) => thread.parentThread)
  threads: Thread[];
}

File: rest/src/main/java/com/waglewagle/rest/thread/Thread.java
package com.waglewagle.rest.thread;

import com.waglewagle.rest.keyword.Keyword;
import com.waglewagle.rest.thread.ThreadDTO.*;
import com.waglewagle.rest.user.User;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class Thread {

    @Id
    @GeneratedValue
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    private User author;

    @Column(length = 5000)
    private String content;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "keyword_id")
    private Keyword keyword;

    @ManyToOne(fetch = FetchType.LAZY)
    private Thread parentThread;

    @OneToMany(mappedBy = "id")
    private List<Thread> children = new ArrayList<>();

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    private LocalDateTime deletedAt;

    public Thread(CreateThreadDTO createThreadDTO) {
        author = createThreadDTO.getAuthor();
        parentThread = createThreadDTO.getParentThread();
        content = createThreadDTO.getContent();
        keyword = createThreadDTO.getKeyword();
    }
}
